updated AzureML deployment method

shivp950 opened the pull request:
## What changes are proposed in this pull request?

Removed the build image method for deploying an MLflow model to AzureML since it is deprecated. I added the new azureml.mlflow.deploy method. 

## How is this patch tested?

Azure ML team developed and validated this method with internal tests 

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Model Deployments to Azure Machine Learning do not require an image build step anymore, there's a new azureml.mlflow.deploy method. 

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ x] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

shivp950 committed:
updated AzureML deployment method

Signed-off-by: Shivani Patel <shipatel@microsoft.com>

shivp950 commented:
@akshaya-a

akshaya-a commented on docs/source/models.rst:
```diff
@@ -677,54 +671,6 @@ accepts the following data formats as input:
     response_json = json.loads(response.text)
     print(response_json)
 
-.. rubric:: Example workflow using the MLflow CLI
```
I don't want to remove these things until the replacement command is in place - wdyt?

shivp950 replied:
yes definitely agree, I'll remove the reference

dbczumar replied:
@akshaya-a I agree that we should keep these CLI docs here until we have a replacement CLI. @shivp950 can you restore this CLI example and indicate that it uses the deprecated `mlflow azureml build-image` CLI & that a replacement is coming soon?

shivp950 commented on docs/source/models.rst:
```diff
@@ -629,21 +629,15 @@ accepts the following data formats as input:
+                                               model_name='mymodelname', 
+                                               service_name='myservice', 
+                                               deployment_config=aci_config, 
+                                               tags=None, mlflow_home=None, synchronous=True)
```
got it, I can remove the defaults

shivp950 commented on docs/source/models.rst:
```diff
@@ -629,21 +629,15 @@ accepts the following data formats as input:
+                                               model_name='mymodelname', 
+                                               service_name='myservice', 
+                                               deployment_config=aci_config, 
+                                               tags=None, mlflow_home=None, synchronous=True)
```
good question, adding a fix

shivp950 committed:
addressed the issues in pr comments

Signed-off-by: Shivani Patel <shipatel@microsoft.com>

shivp950 commented on docs/source/models.rst:
```diff
@@ -629,21 +629,15 @@ accepts the following data formats as input:
-
-    # After the image deployment completes, requests can be posted via HTTP to the new ACI
+    # Register and deploy model to Azure Container Instance (ACI)
+    (webservice,model) = mlflow.azureml.deploy(model_uri='runs:/{}/{}'.format(run.id, model_path),
```
will do

akshaya-a commented on docs/source/models.rst:
```diff
@@ -570,26 +567,6 @@ Example requests:
 For more information about serializing pandas DataFrames, see
 `pandas.DataFrame.to_json <https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_json.html>`_.
 
-The predict command accepts the same input formats. The format is specified as command line arguments.
```
are these changes expected? merge conflict?

dbczumar replied:
@akshaya-a @shivp950 These changes are unrelated to AzureML and should be reverted. They refer to other serving capabilities that MLflow provides and will continue to provide.

akshaya-a commented:
also @dbczumar might have some feedback around which scenarios might be good to document here

dbczumar commented on docs/source/models.rst:
```diff
@@ -628,22 +605,17 @@ accepts the following data formats as input:
+                                               workspace=ws,
+                                               model_name='mymodelname', 
+                                               service_name='myservice', 
+                                               deployment_config=aci_config)
```
@shivp950 This function is synchronous by default (see https://github.com/mlflow/mlflow/blob/master/mlflow/azureml/__init__.py#L215), so the subsequent `webservice.wait_for_deployment()` call is unnecessary.

shivp950 replied:
removed the redundancy

dbczumar commented on docs/source/models.rst:
```diff
@@ -628,22 +605,17 @@ accepts the following data formats as input:
+
+    # Register and deploy model to Azure Container Instance (ACI)
+    (webservice, model) = mlflow.azureml.deploy(model_uri='runs:/{}/{}'.format(run.id, model_path),
+                                               workspace=ws,
+                                               model_name='mymodelname', 
+                                               service_name='myservice', 
+                                               deployment_config=aci_config)
```
Nit: Can we align these arguments with the top line? Looks like the indentation is a bit off

dbczumar commented on docs/source/models.rst:
```diff
@@ -628,22 +605,17 @@ accepts the following data formats as input:
+
+    # Register and deploy model to Azure Container Instance (ACI)
+    (webservice, model) = mlflow.azureml.deploy(model_uri='runs:/{}/{}'.format(run.id, model_path),
+                                               workspace=ws,
```
@shivp950 This variable is undefined in the codeblock; the workspace is referenced by the `azure_workspace` variable

shivp950 replied:
thanks for all the feedback, let me know if you see any additional issues with the latest changes

dbczumar commented on docs/source/models.rst:
```diff
@@ -533,9 +533,6 @@ MLflow can deploy models locally as local REST API endpoints or to directly scor
 MLflow can package models as self-contained Docker images with the REST API endpoint. The image can
 be used to safely deploy the model to various environments such as Kubernetes.
 
-You deploy MLflow model locally or generate a Docker image using the CLI interface to the
```
@shivp950 This also should not be removed.

shivp950 replied:
we haven't updated the cli to use the latest version of model deploy for azureml, if we keep the build image and deploy cli commands, it's using the deprecated aml image.deploy method

dbczumar replied:
@shivp950 The CLI we're referring to in this part of the documentation has nothing to do with AzureML! It should be left as-is!

shivp950 replied:
yes I see now, my bad, I'll fix it

dbczumar commented on docs/source/models.rst:
```diff
@@ -628,22 +605,17 @@ accepts the following data formats as input:
                                        location=location,
```
@shivp950 Can we update the paragraph/description at the beginning of this docs section (`Deploy a python function model on Microsoft Azure ML`) to indicate that the `mlflow.azureml` module provides tools for building deployment images (deprecated) as well as creating new AKS / ACI deployments? This paragraph seems out-of-date and should be updated to reflect the new API.

cc @akshaya-a

shivp950 replied:
makes sense, will do

shivp950 committed:
updated with new feedback

Signed-off-by: Shivani Patel <shipatel@microsoft.com>

dbczumar commented on docs/source/models.rst:
```diff
@@ -595,14 +595,13 @@ For more info, see:
 * JSON-serialized pandas DataFrames in the ``split`` orientation. For example, ``data = pandas_df.to_json(orient='split')``. This format is specified using a ``Content-Type`` request header value of ``application/json``.
 
-* :py:func:`build_image <mlflow.azureml.build_image>` registers an MLflow Model with an existing Azure ML workspace and builds an Azure ML container image for deployment to AKS and ACI. The `Azure ML SDK`_ is required in order to use this function. *The Azure ML SDK requires Python 3. It cannot be installed with earlier versions of Python.*
+* :py:func:`build_image <mlflow.azureml.deploy>` registers an MLflow Model with an existing Azure ML workspace, builds an Azure ML container image and deploys the model to AKS and ACI. The `Azure ML SDK`_ is required in order to use this function. *The Azure ML SDK requires Python 3. It cannot be installed with earlier versions of Python.*
```
@shivp950 I think this should say `deploy()` or `mlflow.azureml.deploy`, rather than `build_image`

```suggestion
* :py:func:`mlflow.azureml.deploy` registers an MLflow Model with an existing Azure ML workspace, builds an Azure ML container image and deploys the model to AKS and ACI. The `Azure ML SDK`_ is required in order to use this function. *The Azure ML SDK requires Python 3. It cannot be installed with earlier versions of Python.*
```

dbczumar commented on docs/source/models.rst:
```diff
@@ -628,22 +627,16 @@ accepts the following data formats as input:
+    aci_config = AciWebservice.deploy_configuration(cpu_cores=1, memory_gb=1)                                       
+
+    # Register and deploy model to Azure Container Instance (ACI)
+    (webservice, model) = mlflow.azureml.deploy(model_uri='runs:/{}/{}'.format(run.id, model_path),
```
`run.id` and `model_path` are not defined in this scope. Can we change this as follows?:

```suggestion
    (webservice, model) = mlflow.azureml.deploy(model_uri='<your-model-uri>',
```

dbczumar reviewed (changes requested):
@shivp950 I left a couple of minor suggestions, but this looks great in general. Should be ready to merge with a couple of minor tweaks.

shivp950 committed:
Update docs/source/models.rst

Co-authored-by: dbczumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Shivani Patel <shipatel@microsoft.com>

shivp950 committed:
updated with sign off to Update docs/source/models.rst

Co-authored-by: dbczumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Shivani Patel <shipatel@microsoft.com>

shivp950 committed:
Update docs/source/models.rst

Co-authored-by: dbczumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Shivani Patel <shipatel@microsoft.com>

shivp950 commented:
@dbczumar I believe I've resolved all the change requests however I'm unable to mark your requests as resolved

shivp950 committed:
updated

force push erased the commit, trying to get the commit changes associated with the pr branch

Commit:
Merge branch 'patch-1' of https://github.com/shivp950/mlflow into patch-1

Signed-off-by: Shivani Patel <shipatel@microsoft.com>
